test/ccall.jl

```diff
@@ -1509,3 +1509,92 @@ let
     @test isa(ptr, Ptr{Cvoid})
     @test arr[1] == '0'
 end
+
+# @ccall macro
+using Base.CcallMacro: @ccall, parsecall, lower, remove_linums, CcallError
+
+@testset "test basic parsecall functionality" begin
+    callexpr = :(
+        libc.printf("%s = %d\n"::Cstring; name::Cstring, value::Cint)::Cvoid
+    )
+    @test parsecall(callexpr) == (
+        :((:printf, libc)),               # function
+        :Cvoid,                           # returntype
+        :((Cstring, Cstring, Cint)),      # argument types
+        Any["%s = %d\n", :name, :value],  # argument symbols
+        1                                 # number of required arguments (for varargs)
+    )
+end
+
+@testset "ensure the base-case of @ccall works, including library name" begin
+    call = lower(:ccall, parsecall( :( libstring.func(
+        str::Cstring,
+        num1::Cint,
+        num2::Cint
+    )::Cstring))...)
+    @test call == remove_linums(quote
+        var"%1" = Base.cconvert($(Expr(:escape, :Cstring)), $(Expr(:escape, :str)))
+        var"%4" = Base.unsafe_convert($(Expr(:escape, :Cstring)), var"%1")
+        var"%2" = Base.cconvert($(Expr(:escape, :Cint)), $(Expr(:escape, :num1)))
+        var"%5" = Base.unsafe_convert($(Expr(:escape, :Cint)), var"%2")
+        var"%3" = Base.cconvert($(Expr(:escape, :Cint)), $(Expr(:escape, :num2)))
+        var"%6" = Base.unsafe_convert($(Expr(:escape, :Cint)), var"%3")
+        $(Expr(
+            :foreigncall,
+            :($(Expr(:escape, :((:func, libstring))))),
+            :($(Expr(:escape, :Cstring))),
+            :($(Expr(:escape, :(Core.svec(Cstring, Cint, Cint))))),
+            0,
+            :(:ccall),
+            Symbol("%4"), Symbol("%5"), Symbol("%6"),
+            Symbol("%1"), Symbol("%2"), Symbol("%3")))
```

Comment:
@vtjnash does `:foreigncall` still need special support for GC roots when we could use `GC.@preserve` for that?

Reply:
Strictly speaking, probably not. But this might generate better code since we get more precise information about the use site.